From 13d64e41f757a3594c4fba3366974200ebfbb6d4 Mon Sep 17 00:00:00 2001
From: Jeff Moyer <jmoyer@redhat.com>
Date: Thu, 28 Aug 2014 19:16:00 -0400
Subject: [virt] virtio_blk, blk-mq: allow blk_mq_init_commands() to return failure

Message-id: <1409253371-23161-11-git-send-email-jmoyer@redhat.com>
Patchwork-id: 89850
O-Subject: [RHEL7 PATCH 10/21] virtio-blk, blk-mq: allow blk_mq_init_commands() to return failure
Bugzilla: 1105698
RH-Acked-by: Fam Zheng <famz@redhat.com>

This is a backport of the virtio-blk portion of the following commit.
This is related to bug 1105698.

  commit 95363efde193079541cb379eb47140e9c4d355d5
  Author: Jens Axboe <axboe@fb.com>
  Date:   Fri Mar 14 10:43:15 2014 -0600

    blk-mq: allow blk_mq_init_commands() to return failure

    If drivers do dynamic allocation in the hardware command init
    path, then we need to be able to handle and return failures.

    And if they do allocations or mappings in the init command path,
    then we need a cleanup function to free up that space at exit
    time. So add blk_mq_free_commands() as the cleanup function.

    This is required for the mtip32xx driver conversion to blk-mq.

    Signed-off-by: Jens Axboe <axboe@fb.com>

Signed-off-by: Jeff Moyer <jmoyer@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/block/virtio_blk.c b/drivers/block/virtio_blk.c
index a2db9ed..0b1c64c 100644
--- a/drivers/block/virtio_blk.c
+++ b/drivers/block/virtio_blk.c
@@ -496,13 +496,14 @@ static struct blk_mq_reg virtio_mq_reg = {
  .flags  = BLK_MQ_F_SHOULD_MERGE,
 };
 
-static void virtblk_init_vbr(void *data, struct blk_mq_hw_ctx *hctx,
+static int virtblk_init_vbr(void *data, struct blk_mq_hw_ctx *hctx,
         struct request *rq, unsigned int nr)
 {
  struct virtio_blk *vblk = data;
  struct virtblk_req *vbr = rq->special;
 
  sg_init_table(vbr->sg, vblk->sg_elems);
+ return 0;
 }
 
 static int virtblk_probe(struct virtio_device *vdev)
-- 
1.7.1